web-js/src/hooks/useTodos_graphql.js

```diff
@@ -0,0 +1,197 @@
+import React from "react";
+import { ApolloClient, gql, HttpLink, InMemoryCache } from "@apollo/client";
+import { useRealmApp } from "../components/RealmApp";
+import {
+  addValueAtIndex,
+  replaceValueAtIndex,
+  updateValueAtIndex,
+  removeValueAtIndex,
+  getTodoIndex,
+} from "../utils";
+import { useWatch } from "./useWatch";
+import { useCollection } from "./useCollection";
+
+function useApolloClient() {
+  const realmApp = useRealmApp();
+  if (!realmApp.currentUser) {
+    throw new Error(`You must be logged in to Realm to call useApolloClient()`);
+  }
+
+  const client = React.useMemo(() => {
+    const graphqlUri = `https://realm.mongodb.com/api/client/v2.0/app/${realmApp.id}/graphql`;
+    // Local apps should use a local URI!
+    // const graphqlUri = `https://us-east-1.aws.stitch.mongodb.com/api/client/v2.0/app/${realmApp.id}/graphql`
+    async function getValidAccessToken() {
+      // An already logged in user's access token might be stale. To guarantee that the token is
+      // valid, we refresh the user's custom data which also refreshes their access token.
+      await realmApp.currentUser.refreshCustomData();
+      return realmApp.currentUser.accessToken;
+    }
+    return new ApolloClient({
+      link: new HttpLink({
+        uri: graphqlUri,
+        // We define a custom fetch handler for the Apollo client that lets us authenticate GraphQL requests.
+        // The function intercepts every Apollo HTTP request and adds an Authorization header with a valid
+        // access token before sending the request.
+        fetch: async (uri, options) => {
+          const accessToken = await getValidAccessToken();
+          options.headers.Authorization = `Bearer ${accessToken}`;
```

Comment:
This (still) hurts my eyes ... we should really provide a package for this.

Reply:
Agreed! It'd be great to have fetch handlers for common GraphQL clients (e.g. [Apollo](https://www.apollographql.com/docs/react/), [urql](https://formidable.com/open-source/urql/docs/), [graphql-hooks](https://github.com/nearform/graphql-hooks)), that can hook into the SDK's built-in access token refresh handler logic.